In [6]:
import pandas as pd
import os
from SPARQLWrapper import SPARQLWrapper, JSON

In [7]:
folderlink = '..//data//'
input_folder = 'input//'
folder_output = 'output//'
ddrc_file = 'parent_data.csv'
input_file = os.path.join(folderlink+folder_output, ddrc_file)


In [8]:
sparql_endpoint = "http://data.bibliotheken.nl/sparql"


In [9]:
# query 1 to get all titles that are available in STCN
sparql_query = """
select ?title_id ?title where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [10]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q1.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           216051   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p82210427X   
freq                                                 1   

                                title  
count                          216051  
unique                         176553  
top     Theses juridicae inaugurales.  
freq                              497

There are 216051 unique book items in STCN of which 176553 are truly unique. Meaning that some book items have the same title or are duplicates. 

In [11]:
# query 2 to get all titles that are available in STCN plus additional data
sparql_query = """
select ?title_id ?title ?fingerprint where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p <http://data.bibliotheken.nl/def#stcnFingerprint> ?fingerprint .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [12]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q2.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           222619   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p375930256   
freq                                                13   

                                title         fingerprint  
count                          222619              214773  
unique                         176553              201960  
top     Theses juridicae inaugurales.  000004 - b1=b2 A e  
freq                              499                 566

There are 216051 unique titles in STCN, there are 201960 fingerprint records. Not every title has a fingerprinte, some titles have multiple fingerprints.

In [13]:
# query 3 to get all titles that are available in STCN plus information about the fingerptint and about the main author
sparql_query = """
select ?title_id ?title ?fingerprint ?author_id where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p <http://data.bibliotheken.nl/def#stcnFingerprint> ?fingerprint .
    ?p schema:author ?author_id .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [14]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q3.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           220080   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p235143030   
freq                                                 9   

                                title fingerprint            author_id  
count                          220080      115660               115660  
unique                         176553      112631               111631  
top     Theses juridicae inaugurales.      000004  nodeID://b266146401  
freq                              499         445                    9

text

In [15]:
# query 4 to get all titles that are available in STCN and information about the main author
sparql_query = """
select ?title_id ?title ?author_id ?author where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:author ?author_id .
    ?p schema:author/schema:name ?author .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [16]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q4.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           216051   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p82220682X   
freq                                                 1   

                                title            author_id       author  
count                          216051               117951       117951  
unique                         176553               117951        51294  
top     Theses juridicae inaugurales.  nodeID://b265835929  De Voltaire  
freq                              497                    1          194

In [34]:
# query 6 to get all titles that are available in STCN and information about the main author and additional information about it
sparql_query = """
select ?title_id ?title ?author_id ?author ?author_givenName where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:author ?author_id .
    ?p schema:author/schema:name ?author .
    ?p schema:author/schema:author/schema:givenName ?author_givenName .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [35]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q6.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           216844   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p832460834   
freq                                                 2   

                                title            author_id         author  \
count                          216844               118730         118730   
unique                         176553               117937          51285   
top     Theses juridicae inaugurales.  nodeID://b266082509  J.J. Rousseau   
freq                              497                    2            202   

       author_givenName  
count            118730  
unique             9802  
top                      
freq               4955

In [36]:
# query 7 to get all titles that are available in STCN and information about the main author and additional information about it. In addition alternative names are added.
sparql_query = """
select ?title_id ?title ?author_id ?author ?author_deathDate ?author_birthDate ?author_familyName ?author_givenName where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:author ?author_id .
    ?p schema:author/schema:name ?author .
    ?p schema:author/schema:author/schema:deathDate ?author_deathDate .
    ?p schema:author/schema:author/schema:birthDate ?author_birthDate .
    ?p schema:author/schema:author/schema:familyName ?author_familyName .
    ?p schema:author/schema:author/schema:givenName ?author_givenName .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [37]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q7.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           218678   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p10217945X   
freq                                                 6   

                                title            author_id         author  \
count                          218678                81900          81900   
unique                         176553                79273          24186   
top     Theses juridicae inaugurales.  nodeID://b266199357  J.J. Rousseau   
freq                              497                    6            404   

       author_deathDate author_birthDate author_familyName author_givenName  
count             81900            81900             81900            81900  
unique             1158             1567              9392             4220  
top                1536             1628                                     
freq               1326             1378              2422             3611

In [42]:
# query 8 to get all titles that are available in STCN and information about the main author and additional information about it. In addition alternative names are added.
sparql_query = """
select ?title_id ?title ?author_id ?author ?author_deathDate ?author_birthDate ?author_familyName ?author_givenName  where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:author ?author_id .
    ?p schema:author/schema:name ?author .
    ?p schema:author/schema:author/schema:deathDate ?author_deathDate .
    ?p schema:author/schema:author/schema:birthDate ?author_birthDate .
    ?p schema:author/schema:author/schema:familyName ?author_familyName .
    ?p schema:author/schema:author/schema:givenName ?author_givenName .
    
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [43]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q8.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           218678   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p10217945X   
freq                                                 6   

                                title            author_id         author  \
count                          218678                81900          81900   
unique                         176553                79273          24186   
top     Theses juridicae inaugurales.  nodeID://b266199357  J.J. Rousseau   
freq                              497                    6            404   

       author_deathDate author_birthDate author_familyName author_givenName  
count             81900            81900             81900            81900  
unique             1158             1567              9392             4220  
top                1536             1628                                     
freq               1326             1378              2422             3611

In [17]:
# query 5 to get all titles that are available in STCN and information about the contributing authors
sparql_query = """
select ?title_id ?title ?contributor_id ?contributor where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:contributor ?contributor_id .
    ?p schema:contributor/schema:name ?contributor .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [18]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q5.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

title_id  \
count                                           251541   
unique                                          216051   
top     http://data.bibliotheken.nl/doc/nbt/p284863009   
freq                                               400   

                                                                                           title  \
count                                                                                     251541   
unique                                                                                    176553   
top     Verhandelingen uitgegeven door het Zeeuwsch Genootschap der Wetenschappen te Vlissingen.   
freq                                                                                        1760   

             contributor_id contributor  
count                 70165       70165  
unique                42161       13316  
top     nodeID://b266206822      Luyken  
freq                     20         792

In [50]:
# query 9 to get all titles that are available in STCN and information about the contributing authors
sparql_query = """
select ?title_id ?title ?contributor_id ?contributor ?contributor_deathDate ?contributor_birthDate ?contributor_familyName ?contributor_givenName where {
  ?p schema:mainEntityOfPage/owl:sameAs ?title_id .
  ?p schema:name ?title .
  OPTIONAL {
    ?p schema:contributor ?contributor_id .
    ?p schema:contributor/schema:name ?contributor .
    ?p schema:contributor/schema:contributor/schema:deathDate ?contributor_deathDate .
    ?p schema:contributor/schema:contributor/schema:birthDate ?contributor_birthDate .
    ?p schema:contributor/schema:contributor/schema:familyName ?contributor_familyName .
    ?p schema:contributor/schema:contributor/schema:givenName ?contributor_givenName .
    }
  ?p schema:mainEntityOfPage/schema:isPartOf <http://data.bibliotheken.nl/id/dataset/stcn> .
} 
"""

In [51]:
sparql = SPARQLWrapper(sparql_endpoint)
sparql.setQuery(sparql_query)
sparql.setReturnFormat(JSON)

results = sparql.query().convert()

bindings = results["results"]["bindings"]
data = []

for item in bindings:
    row = {}
    for key in item:
        row[key] = item[key]["value"]
    data.append(row)

df = pd.DataFrame(data)

# Panda settings for showing data
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)

output_file = folderlink+folder_output+"stcn_q9.csv"
df.to_csv(output_file, sep=';', encoding='utf-8', index=False)

df.describe()

HTTPError: HTTP Error 502: Proxy Error